### PR TITLE
[SPARK-44079][SQL] Fix `ArrayIndexOutOfBoundsException` when parse array as struct using PERMISSIVE mode with corrupt record

### PR DESCRIPTION
**Pull request by Hisoka-X**
### What changes were proposed in this pull request?
When use PERMISSIVE mode to parse json array as structs with `_corrupt_record`. The error will be reported.
```scala
val data = """[{"a": "incorrect", "b": "correct"}, {"a": "incorrect", "b": "correct"}]"""
val schema = new StructType(Array(StructField("a", IntegerType), 
  StructField("b", StringType), StructField("_corrupt_record", StringType)))

spark.read.option("mode", "PERMISSIVE").option("multiline", "true").schema(schema)
  .json(Seq(data).toDS()).show()

// error
org.apache.spark.SparkException: Job aborted due to stage failure: Task 4 in stage 2.0 failed 1 times, most recent failure: Lost task 4.0 in stage 2.0 (TID 9) (charlottesmbp2.home executor driver): java.lang.ArrayIndexOutOfBoundsException: Index 1 out of bounds for length 1
        at org.apache.spark.sql.catalyst.expressions.GenericInternalRow.genericGet(rows.scala:201)
        at org.apache.spark.sql.catalyst.expressions.BaseGenericInternalRow.getAs(rows.scala:35)
        at org.apache.spark.sql.catalyst.expressions.BaseGenericInternalRow.get(rows.scala:37)
        at org.apache.spark.sql.catalyst.expressions.BaseGenericInternalRow.get$(rows.scala:37)
        at org.apache.spark.sql.catalyst.expressions.GenericInternalRow.get(rows.scala:195)
```
The `JacksonParser` wrap row list as a new `InternalRow`. So `FailureSafeParser` can't parse it normally when `_corrupt_record` is defined. 
This PR to make sure `BadRecordException` can cover row list parsed by a string record. Then the `FailureSafeParser` can handle row list one by one.

### Why are the changes needed?
Fix the bug when parse array as struct using PERMISSIVE mode with corrupt record

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
add new test.

### CODE REVIEW

**Comment by Hisoka-X** [comment]
cc @MaxGekk

**Review by MaxGekk — COMMENTED** [review]
@Hisoka-X Thank you for the ping. Could you enable GAs in your fork: https://github.com/apache/spark/pull/41662/checks?check_run_id=14375296234

**Comment by Hisoka-X** [comment]
Hi @MaxGekk ,please check https://github.com/Hisoka-X/spark/actions/runs/5312527795 temporarily. The CI already started, something wrong with github after I forch push code. I will fix this after all CI done.

**Comment by Hisoka-X** [comment]
@MaxGekk Hi, CI passed. Please help me to review. Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/BadRecordException.scala [hunk 1]
```diff
@@ -30,16 +30,29 @@ case class PartialResultException(
...
  * @param partialResult a function that returns an optional row, which is the partial result of
  *                      parsing this bad record.
+ * @param partialResults a function that returns an row array, which is the partial results of
+ *                      parsing this bad record. Only used when the parsing array as structs.
  * @param cause the actual exception about why the record is bad and can't be parsed.
  */
 case class BadRecordException(
     @transient record: () => UTF8String,
     @transient partialResult: () => Option[InternalRow],
+    @transient partialResults: () => Array[InternalRow] = () => Array.empty[InternalRow],
```
[the remark this answers is not in the document]
Comment (Hisoka-X) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/BadRecordException.scala:52]: Thanks for point that. Fixed.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala [hunk 2]
```diff
@@ -319,7 +319,7 @@ class UnivocityParser(
       throw BadRecordException(
         () => getCurrentInput,
         () => None,
-        QueryExecutionErrors.malformedCSVRecordError(""))
+        cause = QueryExecutionErrors.malformedCSVRecordError(""))
```
[the remark this answers is not in the document]
Comment (Hisoka-X) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala]: Done

**Review by MaxGekk — CHANGES_REQUESTED** [review]

**Review by MaxGekk — APPROVED** [review]

**Comment by Hisoka-X** [comment]
Thanks @MaxGekk

**Comment by MaxGekk** [comment]
@Hisoka-X Does 3.4 has the issue? If so, please, open a PR with the backport.

**Review by HyukjinKwon — COMMENTED** [review]
Seems fine to me too

**Pull request closed without merging**